Userland/Libraries/LibJS/Runtime/ExecutionContext.cpp

```diff
@@ -13,9 +13,42 @@
 
 namespace JS {
 
+class ExecutionContextAllocator {
+public:
+    NonnullOwnPtr<ExecutionContext> allocate(Heap& heap)
+    {
+        if (m_execution_contexts.is_empty())
+            return adopt_own(*new ExecutionContext(heap));
+        return m_execution_contexts.take_last();
+    }
+    void deallocate(ExecutionContext* ec)
+    {
+        if (m_destroyed)
+            return;
+        memset(ec, 0, sizeof(ExecutionContext));
+        m_execution_contexts.append(adopt_own(*ec));
```

Comment:
This looks like something UBSAN would not be impressed by.
Why the `memset()`?
I don't think we should be doing `adopt_own(*ec)` on something we've destroyed and memsetted like this.
Perhaps the cache should be `Vector<void*> m_execution_context_slots;` instead, and we just keep the raw malloc pointers there.
Then we can placement-new (`new (slot) ExecutionContext(heap)`) into them.